dreadnode/cli/docker.py

```diff
@@ -0,0 +1,387 @@
+    """
+    Get the specified environment variable from the container and return
+    its value.
+
+    Args:
+        container_name: Name of the container to inspect.
+        var_name: Name of the environment variable to retrieve.
+
+    Returns:
+        str | None: Value of the environment variable, or None if not found.
+    """
+    result = docker_run(
+        ["inspect", "-f", "{{range .Config.Env}}{{println .}}{{end}}", container_name],
+        capture_output=True,
+        timeout=30,
+    )
+    for line in result.stdout.splitlines():
+        if line.startswith(f"{var_name.upper()}="):
+            return line.split("=", 1)[1]
+    return None
+
+
+def docker_login(registry: str, username: str, password: str) -> None:
+    """
+    Log into a Docker registry.
+
+    Args:
+        registry: Registry hostname to log into.
+        username: Username for the registry.
+        password: Password for the registry.
+    """
+    docker_run(
+        ["login", registry, "--username", username, "--password-stdin"],
+        stdin_input=password,
+        capture_output=True,
+        timeout=60,
+    )
+
+
+def docker_ps() -> list[DockerContainer]:
+    """
+    List and parse running containers using `docker ps`.
+
+    Returns:
+        A list of DockerPSResult objects.
+    """
+    result = docker_run(
+        ["ps", "--format", "json"],
+        capture_output=True,
+    )
+    return [
+        DockerContainer.model_validate_json(line)
+        for line in result.stdout.splitlines()
+        if line.strip()
+    ]
+
+
+def docker_compose_ps(args: list[str] | None = None) -> list[DockerContainer]:
+    """
+    List and parse running containers using `docker compose ps`.
+
+    This mirrors:
+        docker compose [*args] ps --format json
+
+    Args:
+        args: Additional docker compose arguments.
+
+    Returns:
+        A list of DockerPSResult objects.
+    """
+    result = docker_run(
+        ["compose", *(args or []), "ps", "--format", "json"],
+        capture_output=True,
+    )
+    return [
+        DockerContainer.model_validate_json(line)
+        for line in result.stdout.splitlines()
+        if line.strip()
+    ]
+
+
+def docker_tag(image: str | DockerImage, new_tag: str) -> None:
+    """
+    Tag a Docker image with a new tag.
+
+    Args:
+        image: The name of the image to tag.
+        new_tag: The new tag to apply to the image.
+    """
+    docker_run(
+        ["tag", str(image), new_tag],
+        capture_output=True,
+        timeout=60,
+    )
+
+
+def get_local_registry_port() -> int:
+    for container in docker_ps():
+        if DEFAULT_DOCKER_REGISTRY_IMAGE_TAG in container.image and container.ports:
+            # return the first mapped port
+            return container.ports[0][0]
+
+    # fallback to the default port
+    return DEFAULT_DOCKER_REGISTRY_LOCAL_PORT
+
+
+def get_registry(config: ServerConfig) -> str:
```

Comment:
Helper code from the original CLI, I'm trusting it generally, but it's probably fragile and needs some more consideration later for on-prem instances when we get there.